compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirUselessTypeOperationCallChecker.kt

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.expression
+
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.expressions.*
+import org.jetbrains.kotlin.fir.resolve.fullyExpandedType
+import org.jetbrains.kotlin.fir.resolve.inference.isFunctionalType
+import org.jetbrains.kotlin.fir.typeContext
+import org.jetbrains.kotlin.fir.types.*
+import org.jetbrains.kotlin.types.AbstractTypeChecker
+
+// See .../types/CastDiagnosticsUtil.kt for counterparts, including isRefinementUseless, isExactTypeCast, isUpcast.
+object FirUselessTypeOperationCallChecker : FirTypeOperatorCallChecker() {
+    override fun check(expression: FirTypeOperatorCall, context: CheckerContext, reporter: DiagnosticReporter) {
+        if (expression.operation !in FirOperation.TYPES) return
+        val arg = expression.argument
+
+        val candidateType = arg.typeRef.coneType.upperBoundIfFlexible().fullyExpandedType(context.session)
+        if (candidateType is ConeKotlinErrorType) return
+
+        val targetType = expression.conversionTypeRef.coneType.fullyExpandedType(context.session)
+        if (targetType is ConeKotlinErrorType) return
+
+        // x as? Type <=> x as Type?
+        val refinedTargetType =
+            if (expression.operation == FirOperation.SAFE_AS) targetType.withNullability(ConeNullability.NULLABLE) else targetType
+        val shouldCheckForExactType = when (expression.operation) {
```

Comment:
In the counterpart in `CastDiagnosticsUtil.kt`, AFAICT I don't see anywhere that looks at the operation (`is` vs `as`). Why the difference?

Reply:
Inside that file, yeah you can't tell. If you search for the usage of `isRefinementUseless`, which is the common entry point for `as` and `is` checks, you will see the difference.

* `PatternMatchingTypingVisitor#detectRedundantIs`, which checks `is`, calls `isRefinementUseless` with `false`, hence `false` constant for `IS` and `NOT_IS` here.

* `CastDiagnosticsUtil#castIsUseless`, which is used for `as`, calls `isRefinementUseless` with `shouldCheckForExactType` result, whose meaning is explained as comments (TODO) below.

Reply:
Oh ok. That was tricky. I thought it was the same entry point (`castIsUseless`) for both.